### ISSUE TITLE
Issues on Argo 2.3

### ISSUE THREAD

**Issue by elikatsis** [kubeflow/pipelines#1250]
Hi everyone!

Argoproj team has cut some new releases for Argo.

I've noticed there will be some incompatibility due to argoproj/argo#1275.
**tl;dr**: A step fails if an artifact path does not exist or points to an empty file.

We could make use of the `optional` field (introduced in argoproj/argo#1277) for the `conventional_artifacts`, which are added by default as output artifacts to every `ContainerOp` (`mlpipeline-ui-metadata`, `mlpipeline-metrics`).
Or, let the user declare if they need these artifacts (e.g. `step.add_ui_metadata()`, `step.add_metrics()`), or the other way around depending on what would be the default behavior.

Maybe that's something to take into consideration for #1003 and #1064, too.

This may, unfortunately, break existing samples if not recompiled (unless there is some other way?).
Also, how would this affect the current usage of these artifacts?

Nevertheless, the update is required to have E2E tests covering #926 and make it work to its full potential, as well as take advantage of any other new feature or bug fix.

If we make use of the `optional` field, I think the kubeflow release updating the images and the compiler release adding `optional: true` should be done _at the same time_.
WDYT?

/cc @Ark-kun @hongye-sun @vicaire

**Comment by hongye-sun** [comment]
We can change the pipeline API to add the optional field for conventional artifacts before sending to Argo To avoid breaking existing pipelines. We might show warning to the user in UI to upgrade their client to recompile.

**Comment by Ark-kun** [comment]
>A step fails if an artifact path does not exist or points to an empty file.

That's pretty bad. Looks like I was right in being strongly against adding those two artifacts to all templates.

I've already done some prerequisite work for resolving the situation, but there were many delays.
There is no no perfect way to fix the issue, bu here is my plan to fix this with least sacrifices:

1. Immediately start marking those two artifacts as optional.
2. Explicitly specify those artifact for the components that really produce them. https://github.com/kubeflow/pipelines/blob/master/components/local/confusion_matrix/component.yaml#L7 Since https://github.com/kubeflow/pipelines/pull/998 it's possible to specify output artifacts in the `ContainerOp`.
3. Change the compiler to stop adding those two artifacts to every op unless they're explicitly specified.

>This may, unfortunately, break existing samples if not recompiled (unless there is some other way?).

If  we were adding the artifact storage configuration in the backend (like I wanted), it would be trivial to start adding the optional flag to them those two. Unfortunately, we do not have this facility at this moment.

**Comment by Ark-kun** [comment]
>If we make use of the optional field, I think the kubeflow release updating the images and the compiler release adding optional: true should be done at the same time.

Argo's and KFP backend use the `json.Unmarshal` parser which ignores unknown keys by default. So we've already proactively started adding `optional=True` both in the compiler and the backend.

We're planning to upgrade to Argo 2.3 once they release the final version and kubeflow upgrades to it.

@elikatsis WDYT?

**Comment by elikatsis** [comment]
You've done a pretty nice job there, @Ark-kun!

Do you think there should be a deprecation warning by the compiler? So that in a couple of releases these artifacts are not automatically added.
Also, should there be a similar UI warning to recompile the pipelines? Because I believe the #1289 hack shouldn't stay there forever.

**Comment by elikatsis** [comment]
Hi everyone!

I have created #1342, a PR adding E2E tests for #926.

Please take a moment to read the rationale behind it, it updates Argo images for the testing infrastructure's deployment.
I consider this would close this issue, since the rest of the goals Alexey has set (stop adding these 2 artifacts and explicitly specify them when they are needed) can be achieved asynchronously, and it requires some time before it is fully accomplished.

Also, I strongly believe this change should be upstream for the next Kubeflow release (https://github.com/kubeflow/pipelines/issues/1327#issuecomment-492995591).

**Comment by Ark-kun** [comment]
Closing this issue as the blocking issues have been fixed by https://github.com/kubeflow/pipelines/issues/1250 and https://github.com/kubeflow/pipelines/issues/1260 and we've already upgraded to Argo 2.3.

**Comment by Ark-kun** [comment]
> in a couple of releases these artifacts are not automatically added.

The compiler no longer auto-adds those artifacts (since v.01.29): https://github.com/kubeflow/pipelines/pull/2046
